main.c

```diff
@@ -435,89 +457,128 @@ inline uint8_t min(uint8_t a, uint8_t b) { return a <= b ? a : b; }
   // overrides the foreground
   if (foregroundColorSet) {
     return;
   }
-
-  profiles[currentProfile].callback(ledColors, ledIntensity);
+  profiles[currentProfile].callback(ledColors);
 }
 
-static inline void sPWM(uint8_t cycle, uint8_t currentCount, ioline_t port) {
-  if (cycle > currentCount) {
-    palSetLine(port);
+/*
+ * Time each row has left to shine within the current column cycle.
+ * Row1 R-G-B, Row2 R-G-B, Row3 R-G-B, ...
+ */
+uint8_t rowTimes[NUM_ROW * 3];
+uint8_t rowsEnabled;
+
+/*
+ * pwmCounter which counts time of lit rows within each column cycle.
+ */
+uint16_t pwmCounter;
+
+/*
+ * pwmCounter goes over 64 a bit to limit the current of completely white
+ * board (0.5A) vs <0.3A for original firmware.
+ *
+ * You can get brighter LEDs if you set this to 64. And possibly burn the
+ * board in the longer period of time.
+ */
+const uint16_t pwmCounterLimit = 80;
+
+/* Disable timeouted LEDs */
+static inline void pwmRowDimmer() {
+  for (size_t ledRow = 0; ledRow < NUM_ROW * 3; ledRow++) {
+    const uint8_t time = rowTimes[ledRow];
+    if (pwmCounter == time) {
+      palClearLine(ledRows[ledRow]);
+      rowsEnabled--;
+    }
+  }
+  if (rowsEnabled == 0) {
+    /* Limit color bleed by disabling the column as early as possible */
+    palClearLine(ledColumns[currentColumn]);
   }
 }
 
-uint8_t rowPWMCount = 0;
+/* Start new PWM cycle */
+static inline void pwmNextColumn() {
+  /* Disable previously lit column */
+  palClearLine(ledColumns[currentColumn]);
+
+  currentColumn = (currentColumn + 1) % NUM_COLUMN;
+
+  /* TODO: Minimum intensity per animation? For some, the darkness doesn't work
+     well (rainbow) */
+  const uint8_t intensityDecrease = ledIntensity * 8;
+
+  /* Prepare the PWM data and enable leds for non-zero colors */
+  rowsEnabled = 0;
+  for (size_t keyRow = 0; keyRow < NUM_ROW; keyRow++) {
+    const uint8_t ledIndex = currentColumn + NUM_COLUMN * keyRow;
+    const led_t cl = ledColors[ledIndex];
+
+    for (size_t colorIdx = 0; colorIdx < 3; colorIdx++) {
+      const uint8_t ledRow = 3 * keyRow + colorIdx;
+      /* >>2 to decrease the color resolution from 0-255 to 0-63 */
+      uint8_t color = cl.pv[2 - colorIdx] >> 2;
+      if (intensityDecrease >= color) {
+        color = 0;
+      } else {
+        color -= intensityDecrease;
+        /* Each led is enabled for color>0 even for a short while. */
+        palSetLine(ledRows[ledRow]);
+        rowsEnabled++;
+      }
+      rowTimes[ledRow] = color;
+    }
+  }
+
+  /* Enable the current LED column if at least one row needs this. Limit bleed
+     and maybe power consumption on reactive profiles. */
+  if (rowsEnabled) {
+    palSetLine(ledColumns[currentColumn]);
+  }
+}
 
 // mainCallback is responsible for 2 things:
 // * software PWM
 // * calling animation callback for animated profiles
 void mainCallback(GPTDriver *_driver) {
   (void)_driver;
 
-  if (ledEnabled) {
+  if (!ledEnabled) {
```

Comment:
Nitpick, this check is probably redundant. When led is disabled, gpt interval is stopped and this function should not be called.

Reply:
This check was there when I started, I've only lessened the indentation there to more easily reason about the logic. It probably is unnecessary, kinda defensive programming practice. It is in the tight loop, so maybe should be removed. 

OTOH, we have time for it. Loop runs @80kHz (12.5µs), and I checked it can reach stabily a bit over 100kHz (every 10µs), so there's at least 2.5µs to spare. Hopefully. That's 120 instructions (@48MHz/48MIPS)?

Reply:
Sounds good, I don't have a strong opinion on this.